### ISSUE TITLE
APSIM version numbers

### ISSUE THREAD

**Issue by femiguez** [APSIMInitiative/ApsimX#6990]
I just downloaded the latest version of APSIM and for Mac it appears as simply APSIM6978 without the date. I see that this was discussed in #6885. Was this intentional? It is not clear to me that it was from that discussion. I'm guessing that it was not as the Windows version appears as 2021.11.6978.0. But in this Windows version, now the 'day' was eliminated. Again, I'm guessing this was unintentional. I'm not convinced that eliminating the date from the version is helpful for most users. If the number is associated with the issue, it would be even more confusing, because the number associated with each version would not correspond with a newer version. Personally, I think this is less clear than the previous versioning with the date stamp. Clarification: the file name has lost the date stamp. The application itself is similar to Windows.

**Comment by hol430** [comment]
As per my second-last comment in #6885, the version numbers now look like year.month.revision.0. The day component has been removed. The revision number is no longer associated with the issue, it's just a unique number which goes up by 1 for each version. These change only apply to new versions.

> I just downloaded the latest version of APSIM and for Mac it appears as simply APSIM6978 without the date.

Are you talking about the upgrades page on the website? The file name of the downloaded installer file? The window title shown in the apsim GUI? Something else?

**Comment by femiguez** [comment]
Sorry! I should have explained myself better. What threw me off is that now in Applications I see the following

<img width="441" alt="MacAPSIMApps" src="https://user-images.githubusercontent.com/10869358/145232693-02f46310-ab8b-4fb0-af73-2c1f525118eb.png">

The date (now just YYYY.MM is gone). And I thought that removing the year, month and date was somewhat confusing. Going back to #6885, my recent experience with teaching and research is that the continuous release cycle ends up being confusing for many users but also problematic in other ways. For example, this is the last week of the semester (thankfully), but in the class I teach we started off with the same version of APSIM. As the semester progressed many students updated APSIM, so by now it feels like every student is running a different version. Because of backward incompatibility, this is problematic. For example, if a student sends me a simulation they are working on and I edit it in a newer version of APSIM and send it back to them they won't be able to open it. In research, we have experienced similar issues. My suggestion is that in the download page the pull down menu would have the options: 'APSIM Classic', 'APSIM Next Gen devel', 'APSIM Next Gen release', 'APSoil'. The release version would have a simpler versioning system than the devel branch and it would only be updated every 6 months. So in essence not too different from @rcichota suggestion. This is my experience and realize that it might not be the same for everyone else. Thanks again!

**Comment by hol353** [comment]
We understand the frustrations of lots of versions but...

Is there any reason why you can't tell the students that the course will be using APSIM Next Gen version 2021.08.12 and to not upgrade? It seems like everyone will want a different release cycle. You want 6 monthly but others will want shorter (or longer) cycles. Instead of different release 'channels', isn't it just a matter of locking in a particular version for the duration of the semester? We do that for project work where we tell all team member to use version xyz. We then might change that during the project but that is the decision of the project team. Would this work?

**Comment by jbrider** [comment]
@femiguez The recent changes are to largely address the issues you mention here.
I think what has happened is that the year and month part have temporarily disappeared from the filename.
The last part of the name used to be a reference to the github issue - but was causing issues as the issue numbers was not necessarily mathematically ordered - which is being changed to a continuously updated number as part of the minor release part of the version naming (I think that's the right name for the last part of it?).  The day was then removed to make the version yyyy.mm.vvvv.
The other change that was made was to the downloads page so people could link to a specific version - so you could publish a specific link to the version you wanted all your students to use.

**Comment by femiguez** [comment]
@jbrider @hol353 thanks for your input! Students did upgrade even when I asked them not too. :( This is not a huge deal, but it made things a bit complicated. I was providing an example of the kind of confusion that occurs with making the continuous release cycle available to most users. However, here I wanted to point out that the YYYY.mm have disappeared from the file name for Mac. At first, I was more concerned about this, because if it referenced the issue as it did before, then the version number would not correspond to a newer version of APSIM. If the number will strictly increase, this is less problematic, but the YYYY.mm has not disappeared in Windows, only for Mac, so I thought that may be this was unintentional.

**Comment by hol430** [comment]
I didn't notice that the installation directories on windows still contain the date. I note that the revision number is enough to uniquely identify the build by itself, but I'm happy to add the date component back in on mac/linux if you think that would be useful to have (as opposed to removing the date from the directory name on windows).

**Comment by lie112** [comment]
I have had a quick think of potential ideas
- A agree that the current no forward compatibility ie newer sim than current version, is the cleanest/safest approach, but it can lead trouble working with others and students when sharing simulation files. Being in the continuous upgrade process needed to get the updated code to users as quick as possible I need the most rapid upgrade cycle,
- Is the issue just alerting the user that there have been model changes in the simulation file version that may not work in your installed version and opening the file may delete information needed in the later version? 
- A lock switch could be added on upgrade/downgrade section that writes a file (or just provide the file to students) with the lock version to APSIM directory that locks the model at the given upgrade. This would prevent any student upgrades until the file is deleted (somewhat messy). It's a bit clunky but may serve the purpose. Just thinking about some more manual bespoke approaches that may allow groups using APSIM to control upgrades and versioning. It could be that the annual upgrades can have a setting to lock to annual upgrade.
- Another option would be able to set the max version that a simulation file can be opened in. This could be set by lecturers when handing out the example simulations, but wouldn't help when created form scratch. 
- Most of the time upgrades are backward compatible, but it is not possible to determine when they are not or whether there is any change in logic and output with each upgrade release that would give two students different results based on version used. Therefore APSIM assumes no backward compatibility and thus locks based on version and gives the error and warning if a newer simulation file is opened in an old software version.  There is a version recorded in the apsimx simulation file, could this be used or have another property that restricts the version that the file can be opened with. rather than no backward compatibility (Or just a big warning that can be accepted or ignored when trying to open newer sim file in an old version), could there also be a max version that the file can be opened with, thus restricting the versions a student can use and allowing the lecturer to open and use a student version if possible. 
-  I have wondered whether there is a more automated way of allowing apsim files to be opened in any version and reporting a useful error message if the simulation file cannot be opened?

**Comment by femiguez** [comment]
@hol430 I don't know what naming convention for the files is more clear or intuitive. My view is that since the distributed binaries had the date stamp for years now, I was expecting to see it there still. I realize this is subjective. However, consistency among platforms is desirable. My preference is to keep at least YYYY.mm for all platforms, but for a new user having a simpler version system just based on a single number might be more intuitive. I welcome comments from others about this.

I do not feel too strong about the issue of backwards compatibility. I mentioned it as, what it seems to me, an example of the more complex issue of having too many 'versions' of APSIM out there. It is inconvenient, but from a software development point of view, it probably avoids many other headaches. Overall, I learned to adapt to this and really was not too frustrated.

As a result of this discussion, I am planning to distribute a survey for students in which I am asking specifically about their APSIM experience. One hypothesis is that since upgrading is very easy (which is great!) when they encounter an error, there is the temptation of upgrading APSIM to see if the error goes away. There are two pieces to this. 1) Even when the error message is crystal clear, the lack of experience results in confusion and then they resort to various ways of "fixing" it, like upgrading, even when this has nothing to do with the root cause of the error. 2) Generating better error message is an important (but perhaps under-appreciated?) part of software development. In my packages, I frequently encounter users doing things that I did not expect and getting obscure error messages as a result. When I can, I try to include better error handling, but it does feel like an endless task.

I am very grateful for the work everyone is doing and my comments are not intended to suggest development priorities, just feedback from my experience. I will report back from the survey after it is completed and see if I learn something.
